### DISCUSSION TITLE
How to register custom media handlers with only some resources?

### DISCUSSION THREAD

**Discussion by dkbarn** [falconry/falcon#1924] [Q&A]
What is the correct way to support a specific Content-Type for only a subset of resources?

Example: I have a falcon App with multiple resources/endpoints, some are GET only and return health check statuses, etc. Some of the endpoints allow GraphQL queries via POST. For these resources, I want to register a custom media handler for the "application/graphql" content type.

It seems that media handlers are assigned to the RequestOptions of the application as a whole. But it doesn't make sense for all of my endpoints to receive GraphQL queries. What is the recommended way of handling this scenario?

**Comment by vytas7** [comment]
Hi @dkbarn!
I don't think there is idiomatic support for that out of the box.

One way could be simply adding all handlers, and checking [`req.content_type`](https://falcon.readthedocs.io/en/stable/api/request_and_response_wsgi.html#falcon.Request.content_type) in [hooks](https://falcon.readthedocs.io/en/stable/api/hooks.html) applied only to a specific subset of resources/responders, or doing the same thing in [middleware](https://falcon.readthedocs.io/en/stable/api/middleware.html) (in `process_resource`).

See, for instance, this type of check in our [Quickstart](https://falcon.readthedocs.io/en/stable/user/quickstart.html#a-more-complex-example) docs (performed in middleware):
```python
        if req.method in ('POST', 'PUT'):
            if 'application/json' not in req.content_type:
                raise falcon.HTTPUnsupportedMediaType(
                    title='This API only supports requests encoded as JSON.',
                    href='http://docs.examples.com/api/json')
```

Furthermore, maybe it's not a big deal to have all media handlers registered for the whole app?
When it comes to request media parsing, Falcon does nothing automagical behind your back unless you actually call [`req.get_media()`](https://falcon.readthedocs.io/en/stable/api/request_and_response_wsgi.html#falcon.Request.get_media) (or reference the [`req.media`](https://falcon.readthedocs.io/en/stable/api/request_and_response_wsgi.html#falcon.Request.media) property). So those `GET` health check responders shouldn't be affected in any way.

If the above doesn't fit your bill, we could file a new enhancement proposal issue.
Normally, I would think global handlers are more than enough for small and medium sized (micro)services, but if you are pooling lots of different functionality in a large Falcon app, this functionality might actually make sense.
